timemachine/fe/dummy.py

```diff
@@ -581,3 +612,171 @@ def generate_optimal_dg_ag_pairs(mol, core, bond_idxs):
+    torsion_idxs,
+    torsion_params,
+    min_bond_k=10.0,
+    min_bond_l=0.02,
+    min_angle_k=10.0,
+    min_angle_offset=0.05,
+):
+    """
+    Prune extraneous dummy-core ixns that may be numerically unstable. Sources of numerical instability typically
+    are due to either missing interactions, or incompatible parameters. This does not prune interactions in the
+    dummy-dummy regions since they are required for correctness. While core-core interactions can be pruned, we
+    choose to leave them untouched for now.
+
+    0) All bond terms are untouched.
+    1) All angles i-j-k must have bond terms i-j and j-k present, and with k > 10 and b > 0.2 angstrom on said bonds
+    2) All torsions i-j-k-l must have angle terms i-j-k and j-k-l present, and with k > 10 and
+        abs(angle - pi) > 0.05 rad and abs(angle - 0) > 0.05. In addition, bonded terms i-j, j-k,
+        and k-l must be present, and with k > 10 and b > 0.2 angstrom
+
+    Note that this does not *strictly* guarantee that the system is stable. For example, ring systems do not have
+    angles/bonds set to their equilibrium values:
+
+                D
+               /
+        0--1--2
+        |  |  |
+        3--4--5
+
+    The torsion term 0-1-2-D is numerically unstable, even though the angles and bonds are well-defined. Note that the
+    0-1-2 angle term is not defined in the forcefield as having an equilibrium angle of zero. One additional sanity check
+    that we can probably add later on is to also check the *initial* geometry's bond/angle length parameters.
+
+    Conversely, an interaction may still be stable even if certain sub-component interactions are missing. For example,
+    improper torsions are missing bonds and angles in the trefoil definitions, but the geometry is well defined nevertheless.
+
+    Parameters
+    ----------
+    core: list of int
+        core indices
+
+    bond_idxs: list of 2-tuples
+        (i,j) tuples of the bonded terms
+
+    bond_params: list of 2-tuples
+        (k, b) force constant, bond length
+
+    angle_idxs: list of 3-tuples
+        (i,j,k) tuples where the angle is defined between i-j-k
+
+    angle_params: list of 2-tuples
+        (k, a) force constant, angle in radians
+
+    torsion_idxs: list of 4-tuples
+        (i,j,k,l) tuples where the torsion is defined over i-j-k-l
+
+    torsion_idxs: list of 3-tuples
+        (k, p, n) force constant, period, phase
+
+    Returns
+    -------
+    keep_angle_flags, keep_torsion_flags
+        boolean flags indicating which angles and torsions we keep
+
+    """
+
+    assert len(bond_idxs) == len(bond_params)
+    assert len(angle_idxs) == len(angle_params)
+    assert len(torsion_idxs) == len(torsion_params)
+
+    bond_kv = {}
+    for ij, params in zip(bond_idxs, bond_params):
+        bond_kv[ordered_tuple(ij)] = params
+
+    keep_angle_flags = []
+
+    def ixn_is_all_core_or_all_dummy(atom_idxs):
+        all_core = np.all([a in core for a in atom_idxs])
```

Comment:
nit: would it make sense to use builtin `all` here with generator comprehension? E.g.
```suggestion
        all_core = all(a in core for a in atom_idxs)
```

Reply:
I honestly haven't used python's `all` enough - I'm more familiar with `np.all` (esp when I need to specify axes etc, though it doesn't matter here.

Reply:
(adding this inline directly since I rebased)